beetmoverscript/src/beetmoverscript/script.py

```diff
@@ -285,6 +290,208 @@ async def upload_data(context):
+                 },
+                 "/path/to/public/logs/live.log": {
+                     "destinations": [
+                         "logdest/live.log",
+                     ]
+                 },
+             }
+         }
+     ]
+
+     Each artifact may match up to two `paths` in the `artifactMap`: a `*` path
+     and one other path. If an artifact matches both of these, the non-`*` one
+     takes precedence, and it will _not_ be written to the `*` path destinations.
+     This allows us to have specific filenames (or extensions, prefixes, etc.)
+     grouped together, and to have a "catch-all" for any unmatched artifacts.
+
+     Sanity checking is done to ensure that no destination files would overwrite
+     each other.
+    """
+
+    concreteArtifactMap = []
+    errors = []
+
+    for map_ in artifactMap:
+        concretePaths = {}
+
+        full_glob_destinations = map_["paths"].get("*", {}).get("destinations")
+        other_paths = copy.deepcopy(map_["paths"])
+        if "*" in other_paths:
+            del other_paths["*"]
+
+        for taskId, artifacts in upstreamArtifactPaths.items():
+            if map_["taskId"] != taskId:
+                continue
+
+            for artifact in artifacts:
+                dest_artifact = remove_segments(artifact, strip_segments)
+
+                destinations = []
+                # We need to look at non-'*' paths separate from '*' paths.
+
+                for input_path, output in other_paths.items():
+                    # Skip any input paths that don't match the artifact name.
+                    if "*" in input_path:
+                        if not fnmatch.fnmatch(dest_artifact, input_path):
+                            continue
+                    else:
+                        if input_path != artifact:
+                            continue
+
+                    # If there's already destinations, we've already seen this artifact,
+                    # and we have a clash.
+                    if destinations:
+                        errors.append(f"'{artifact}' matched multiple concrete paths")
+                    else:
+                        destinations.extend(output["destinations"])
+
+                # If we didn't find any destinations for the artifact above, and
+                # there are full glob destinations, it should go there.
+                if full_glob_destinations and not destinations:
+                    destinations.extend(full_glob_destinations)
+
+                # If there are destinations, create a fully concrete entry,
+                # mapping the full path to the file on disk to destinations
+                # it belongs, with any requested segments removed.
+                if destinations:
+                    concretePaths[artifact] = {"destinations": []}
+                    for d in destinations:
+                        if not d.endswith(dest_artifact):
+                            d = os.path.normpath(f"{d}{dest_artifact}")
```

Comment:
Should this require that `d` ends in `/`?  Or join the paths?  Or are there cases where the paths do need to be glued together like that?